### PR TITLE
Fix remaining issues reporting undefined symbols in MAIN_MODULE/RELOCATABLE/LINKABLE builds

### PR DESCRIPTION
**Pull request by sbc100**
Move `gcc_personality_v0.c` to `libunwind`.  This prevents this object file
from being included except when libunwind is present.  Without this, including
`libcompiler-rt` with `-whole-archive` produces undefined symbols for 
`_Unwind_GetIP` and other `libunwind` symbols.

Also, allow programs built with `-sLINKABLE` to report undefined symbol
errors.  By default we don't report errors, but this can be overridden on the
command line).  The issue here was that `MAIN_MODULE=1` sets 
`LINKABLE` which meant we were then not actually reporting  any undefined
symbols.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -1442,14 +1442,18 @@ def default_setting(name, new_default):
         settings.EXPORT_ALL = 1
       settings.RELOCATABLE = 1
 
+    if settings.MAIN_MODULE:
+      settings.DEFAULT_LIBRARY_FUNCS_TO_INCLUDE += [
+          '$getDylinkMetadata',
+          '__heap_base',
+          '__stack_pointer',
+      ]
```
Comment (kripken) [inline emcc.py:1450]: why are these only needed in a main module?

**Review by kripken — COMMENTED** [review]
> Also, allow programs built with -sLINKABLE to report undefined symbol errors. By default they don't (but this can be overridden on the command line).

Where is that default set?

File: emcc.py (hunk 1 above)
Comment (sbc100) [reply emcc.py]: `getDylinkMetadata` (in `library_dylink.py`) is only defined when `MAIN_MODULE` is defined.  Without this change I see `$getDylinkMetadata` as undefined when building with `RELOCATABLE`.

The other two I was able to move back to the original location.

Comment (sbc100) [reply emcc.py]: > > Also, allow programs built with -sLINKABLE to report undefined symbol errors. By default they don't (but this can be overridden on the command line).
> 
> Where is that default set?

Ah I think I meant `RELOCATABLE`:

```
if settings.RELOCATABLE:                                                                         
      default_setting('ERROR_ON_UNDEFINED_SYMBOLS', 0)                                                          
      default_setting('WARN_ON_UNDEFINED_SYMBOLS', 0)    
```

I guess I need to update that block to include LINKABLE too.

**Comment by sbc100** [comment]
I update the PR title and description to reflect what this change is really motivated by.

**Comment by aheejin** [comment]
We don't use gcc_personality_v0.c, and it overlaps with [libcxxabi/src/cxa_personality.cpp](https://github.com/emscripten-core/emscripten/blob/main/system/lib/libcxxabi/src/cxa_personality.cpp), even though they don't look exactly the same. I think we don't need to include this file in libunwind.

File: tools/system_libs.py [hunk 2]
```diff
@@ -961,15 +962,19 @@ class libcxx(NoExceptLibrary, MTLibrary):
...
   includes = [['system', 'lib', 'libunwind', 'include']]
-  src_files = ['Unwind-wasm.cpp']
 
   def __init__(self, **kwargs):
     super(libunwind, self).__init__(**kwargs)
 
+  def get_files(self):
+    return [
+      shared.path_from_root('system', 'lib', 'libunwind', 'src', 'Unwind-wasm.cpp'),
+      shared.path_from_root('system', 'lib', 'compiler-rt', 'lib', 'builtins', 'gcc_personality_v0.c'),
```
Comment (aheejin) [inline tools/system_libs.py:975]: We don't use gcc_personality_v0.c, and it overlaps with [libcxxabi/src/cxa_personality.cpp](https://github.com/emscripten-core/emscripten/blob/main/system/lib/libcxxabi/src/cxa_personality.cpp), even though they don't look exactly the same. I think we don't need to include this file in libunwind.

File: system/lib/pthread/library_pthread_stub.c [hunk 3]
```diff
@@ -139,14 +139,16 @@ int pthread_setspecific(pthread_key_t key, const void* value) {
...
-int pthread_once(pthread_once_t* once_control, void (*init_routine)(void)) {
+int __pthread_once(pthread_once_t* once_control, void (*init_routine)(void)) {
   if (*once_control != PTHREAD_ONCE_MAGIC_ID) {
     init_routine();
     *once_control = PTHREAD_ONCE_MAGIC_ID;
   }
   return 0;
 }
 
+weak_alias(__pthread_once, pthread_once);
```
Comment (kripken) [inline system/lib/pthread/library_pthread_stub.c:150]: what's going on with `pthread_once` here?

Comment (sbc100) [reply system/lib/pthread/library_pthread_stub.c:150]: If you include libc with `--whole-archive` it ends up calling `__pthread_once` (from system/lib/libc/musl/src/thread/call_once.c).   The only time we were doing `--whole-archive` in practice was with `MAIN_MODULE=1` and we were not reported errors on undefined symbols before. (the LINKABLE setting was causing all error to be suppressed, now it isn't).

File: tools/system_libs.py (hunk 2 above)
Comment (sbc100) [reply tools/system_libs.py:975]: OK, maybe I'll try just removing it completely.

File: tools/system_libs.py [hunk 4]
```diff
@@ -961,7 +962,6 @@ class libcxx(NoExceptLibrary, MTLibrary):
 class libunwind(NoExceptLibrary, MTLibrary):
   name = 'libunwind'
   cflags = ['-Oz', '-D_LIBUNWIND_DISABLE_VISIBILITY_ANNOTATIONS']
-  src_dir = ['system', 'lib', 'libunwind', 'src']
```
Comment (aheejin) [inline tools/system_libs.py:964]: Do we not need this anymore? Can we specify where Unwind-wasm.c is without this?

Comment (sbc100) [reply tools/system_libs.py]: Good catch yes, that was not indended.

File: emcc.py (hunk 1 above)
Comment (sbc100) [reply emcc.py]: Done

File: tools/system_libs.py [hunk 5]
```diff
@@ -690,7 +691,7 @@ def get_files(self):
         'getgrouplist.c', 'initgroups.c', 'wordexp.c', 'timer_create.c',
         'faccessat.c',
         # 'process' exclusion
-        'fork.c', 'vfork.c', 'posix_spawn.c', 'execve.c', 'waitid.c', 'system.c'
+        'fork.c', 'vfork.c', 'posix_spawn.c', 'posix_spawnp.c', 'execve.c', 'waitid.c', 'system.c'
```
Comment (kripken) [inline tools/system_libs.py:694]: is this addition a drive-by fix, or related to the PR?

**Review by kripken — APPROVED** [review]

File: tools/system_libs.py (hunk 5 above)
Comment (sbc100) [reply tools/system_libs.py:694]: Its required.. this file contains an umimplementd syscall (same as posix_spawn.c) which results in an undefined symbols if you force include the whole of libc (which is what MAIN_MODULE=1 does).
